### ISSUE TITLE
Refactor Memory plugin for Jaunch

### ISSUE THREAD

**Issue by hinerm** [fiji/fiji#351]
The [Memory plugin](https://github.com/fiji/fiji/blob/a624ffe6c8170c2ca76a7102ec220dd3539e2eb1/src/main/java/sc/fiji/compat/Memory.java) should be updated to read/write key:value pairs.

Relevant classes:
* [OptionsSynchronizer](https://github.com/imagej/imagej-legacy/blob/7eac77ab28901975fc6bf0566a899cf1bef06d97/src/main/java/net/imagej/legacy/OptionsSynchronizer.java)
* [MenuRefresher](https://github.com/fiji/fiji/blob/a624ffe6c8170c2ca76a7102ec220dd3539e2eb1/src/main/java/sc/fiji/compat/MenuRefresher.java)

See also https://github.com/fiji/fiji/issues/350

**Comment by ctrueden** [comment]
It might also be necessary to tweak the `fiji.toml` Jaunch configuration to grab the memory setting out of `fiji.cfg`. In theory it's as simple as writing `${jvm.max-heap}` in the right place, but Jaunch's `jvm.max-heap` key might not support variable interpolation, I can't remember. Testing needed. See also apposed/jaunch#53 and apposed/jaunch#7.

**Comment by hinerm** [comment]
>might not support variable interpolation, I can't remember. 

Interpolation seems to work fine for `fiji.toml` pulling a setting out of `fiji.cfg`, though I haven't figured out how to set a base value if it's unset in `fiji.cfg`.

In https://github.com/apposed/jaunch/issues/53 we discussed modifying `ImageJ.cfg` but I'm not sure I like the idea of that adaptation. I didn't realize that on Mac we don't have an `ImageJ.cfg`  - so any change feels somewhat intrusive (either adding `ImageJ.cfg` on mac and repurposing it on windows, or just moving to a new config file).

Since we're already using `fiji.cfg` for pointing to the JVM location I'd like to just put the max heap setting there.

So I think the changes needed would be:
- [ ] Update `fiji.toml` to read a `jvm.max-heap` value from `fiji.cfg`
- [ ] Keep using `75%` as a default if unset in `fiji.cfg`
- [ ] Modify `Memory and Threads` to set `jvm.max-heap` in `fiji.cfg` on all platforms
- [ ] I'm not sure how to handle *existing* `ImageJ.cfg` files? On Windows they take precedence even when launching with Jaunch, and I don't know why

Thoughts on the last point? Does this sound reasonable or am I missing something?

**Comment by ctrueden** [comment]
Yes, I would prefer to ignore ImageJ.cfg altogether on all platforms, and use fiji.cfg on all of them. :+1:

> I'm not sure how to handle existing ImageJ.cfg files? On Windows they take precedence even when launching with Jaunch, and I don't know why

Whoa, that's surprising! Let's get to the bottom of that.

**Comment by hinerm** [comment]
> Whoa, that's surprising! Let's get to the bottom of that.

🙄🙄 because when you click the taskbar to GC and display max memory, *of course* it [uses what's set in the cfg](https://github.com/imagej/ImageJ/blob/7257bdd07f4ed7ca4703fcc15109d493e597757c/ij/IJ.java#L953) and not the **actual** value.

If you run a script or something to print `Runtime.maxMemory` it will print the correct amount. So, that's something else that probably needs patching in the legacy layer.

**Comment by hinerm** [comment]
OK, so what currently happens is that we interpolate program arguments [here](https://github.com/apposed/jaunch/blob/5f573152532946120191c9598de3b0e086b47393/src/commonMain/kotlin/main.kt#L107). Our `-Xmx${cfg.max-heap}` is properly interpolated. However, the base `config.jvmMaxHeap` field is *never* interpolated.

Unfortuantely, at that point we've already [configured runtimes](https://github.com/apposed/jaunch/blob/5f573152532946120191c9598de3b0e086b47393/src/commonMain/kotlin/main.kt#L91), adding the un-interpolated `config.jvmMaxHeap` as an argument to be interpolated later. This is fine if the memory setting is of the standard type supported by `Xmx`, but *not* if it's a `%`.

Two possible ways forward:
* We could interpolate the actual `JaunchConfig` fields after `Vars` is established. This ensures they can be used freely everywhere without concern of interpolation later. It also feels a little redundant since I assume we still need to interpolate program/runtime args.
* The post-processing of the `jvm` runtime to resolve `%`-based memory specifications is something that could be done in a dedicated argument processing step, after interpolation. This feels slightly better to me: after interpolating args we can just ask the runtime "is there any more custom work that needs to be done with your arguments?" and that would formalize a hook for the % memory expansion, or other activities.

**Comment by ctrueden** [comment]
I like your idea of an additional "last chance" argument-adjustment hook, which the jvm runtime implementation can leverage to fix any `%` present in `-Xmx` and/or `-Xms` args! :+1:
